### PR TITLE
 Initial preparation of common as separate npm package which includes:

### PR DESCRIPTION
**Pull request by rosen-vladimirov**
## Initial preparation of common as separate npm package which includes:
* Add package.json
* Update reference paths to .d.ts
* Add Gruntfile.js

## Provide a way to expose Promisified public API

Use Typescript decorators in order to provide an easy way to expose public API.
Created two decorators:
* `register` - class decorator that can be used instead of `$injector.register`. Usage is `@decorators.register(<module_name>)`
* `promisify` - creates a wrapper around method in order to return promise. The method itself is kept untouched, but the wrapped is added to the publicApi object. Usage is `@decorators.promisify(<module_name>)`

Add requirePublic method in yok in order to create public api. When a module has some methods that must be added to public API, it has to be required with "requirePublic".

Set `getFileSize` of fs module to be promisified in order to allow testing.

### CODE REVIEW

**Comment by rosen-vladimirov** [comment]
In order to test the code, you should:
* Get this branch
* Execute `npm install` in the directory
* Execute `grunt pack`
* Get the produced tgz and install it somewhere `npm i mobile-cli-lib-0.0.1-undefined.tgz` - this way it will be placed in node_modules.
* Use some JS, for example:

```JavaScript:
var common = require("mobile-cli-lib");
common.fs.getFileSize("D:\\Work\\t.txt")
    .then(function (a) {
    console.log("in first then");
    console.log(a);
    return a;
})
    .then(function (a) {
    console.log("in second then");
    console.log(a);
})
    .catch(function (err) {
    console.log("in catch");
    console.log(err);
    console.log("stack = ");
    console.log(err.stack);
});
```

File: definitions/yok.d.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 interface IInjector extends IDisposable {
 	require(name: string, file: string): void;
 	require(names: string[], file: string): void;
+	requirePublic(name: string, file: string): void;
+	requirePublic(names: string[], file: string): void;
```
Comment (ligaz) [inline definitions/yok.d.ts]: Can a union type be used here `string | string[]`?

File: file-system.ts [hunk 2]
```diff
@@ -7,7 +6,9 @@ import path = require("path");
 import util = require("util");
 import rimraf = require("rimraf");
 import minimatch = require("minimatch");
+import decorators = require("./decorators");
 
+@decorators.register("fs")
```
Comment (ligaz) [inline file-system.ts]: From a pure API perspective I think it will be better to move the decorators into the injector file itself. It will look like:
```ts
@injector.register("fs")
```

File: file-system.ts [hunk 3]
```diff
@@ -132,7 +133,8 @@ export class FileSystem implements IFileSystem {
 
 		return future;
 	}
-
+	
+	@decorators.promisify("fs")
```
Comment (ligaz) [inline file-system.ts]: Is the parameter `"fs"` really needed here? Can't it be resolved from the class level decorator?

File: yok.ts [hunk 4]
```diff
@@ -29,7 +29,7 @@ let FN_NAME_AND_ARGS = /^function\s*([^\(]*)\(\s*([^\)]*)\)/m;
 let FN_ARG_SPLIT = /,/;
 let FN_ARG = /^\s*(_?)(\S+?)\1\s*$/;
 let STRIP_COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/mg;
-
+var Promise = require("bluebird");
```
Comment (ligaz) [inline yok.ts]: Is this needed?

File: yok.ts [hunk 5]
```diff
@@ -137,9 +137,29 @@ export class Yok implements IInjector {
...
+			this.requireOne(name, file);
+			this.resolvePublicApi(name, file);
+		});
+	}
+
+	private resolvePublicApi(name: string, file: string): void {
+		Object.defineProperty(this.publicApi, name, {
+			get: () => {
+				let obj: any = {};
+				let originalModule = this.resolve(name);
```
Comment (ligaz) [inline yok.ts]: Is something missing here? I cannot see where we populate the `_publicApi` field.

File: Gruntfile.js [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+	grunt.initConfig({
+		jobName: process.env["JOB_NAME"] || defaultEnvironment,
+		buildNumber: process.env["BUILD_NUMBER"] || "non-ci",
+		dateString: now.substr(0, now.indexOf("T")),
+
+		pkg: grunt.file.readJSON("package.json"),
+
+		ts: {
+			options: {
```
Comment (ligaz) [inline Gruntfile.js]: Do we need those options as there is `tsconfig.json` file as well?

File: yok.ts (hunk 5 above)
Comment (rosen-vladimirov) [inline yok.ts]: when this.resolve is called, the injector requires the file. During this process, the decorators are executed and in `promisify` decorator we populate _publicApi.

File: file-system.ts (hunk 3 above)
Comment (rosen-vladimirov) [inline file-system.ts]: I couldn't find a way to skip the "fs" parameter. The problem is that method decorators are executed before class decorators, so at the moment when the method is promisified, our injector still does not know anything about "fs"

File: yok.ts (hunk 5 above)
Comment (ligaz) [inline yok.ts]: Indeed, I saw this after I posted my comment. Having that in mind I would suggest to rename the `promisify` decorator to `public` or `export` to better donate what is is doing under the hood.

File: definitions/yok.d.ts [hunk 7]
```diff
@@ -24,6 +26,8 @@ interface IInjector extends IDisposable {
 	isValidHierarchicalCommand(commandName: string, commandArguments: string[]): boolean;
 	getChildrenCommandsNames(commandName: string): string[];
 	buildHierarchicalCommand(parentCommandName: string, commandLineArguments: string[]): any;
+	publicApi: any;
+	_publicApi: any;
```
Comment (Fatme) [inline definitions/yok.d.ts]: Do we need this to be part of interface?

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (Fatme) [inline tsconfig.json]: Is this file needed?

File: Gruntfile.js [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+				}
+			},
+
+			build_package: {
+				command: "npm pack",
+				options: {
+					execOptions: {
+						env: (function() {
+							var env = shallowCopy(process.env);
+							env["APPBUILDER_SKIP_POSTINSTALL_TASKS"] = "1";
```
Comment (Fatme) [inline Gruntfile.js]: Is this name correct `APPBUILDER_SKIP_POSTINSTALL_TASKS` ?

File: Gruntfile.js [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+							return env;
+						})()
+					}
+				}
+			}
+		},
+
+		clean: {
+			src: ["test/**/*.js*", "**/*.js*", "!vendor/*.js", "!hooks/**/*.js", "*.tgz"]
```
Comment (Fatme) [inline Gruntfile.js]: `hooks` folder is not needed here

File: common-lib.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+///<reference path=".d.ts"/>
+
```
Comment (teobugslayer) [inline common-lib.ts]: `"use strict";`

File: decorators.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
+///<reference path=".d.ts"/>
+
```
Comment (teobugslayer) [inline decorators.ts]: `"use strict";`

File: fiber-bootstrap.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import Fiber = require("fibers");
```
Comment (teobugslayer) [inline fiber-bootstrap.ts]: `"use strict";`

File: tsconfig.json (hunk 8 above)
Comment (rosen-vladimirov) [inline tsconfig.json]: in fact no

Comment (teobugslayer) [inline tsconfig.json]: I would prefer to introduce tsconfig and remove the code from the gruntfile

File: fiber-bootstrap.ts (hunk 13 above)
Comment (rosen-vladimirov) [inline fiber-bootstrap.ts]: done

File: definitions/yok.d.ts (hunk 7 above)
Comment (rosen-vladimirov) [inline definitions/yok.d.ts]: yes, it is filled in the decorator, so I need it public

File: Gruntfile.js (hunk 6 above)
Comment (rosen-vladimirov) [inline Gruntfile.js]: yes, tsconfig.json is not respected from grunt :(

Comment (ligaz) [inline Gruntfile.js]: Crap, we should look for alternatives :smile:

File: definitions/yok.d.ts (hunk 7 above)
Comment (ligaz) [inline definitions/yok.d.ts]: Can we just use a private field in the publicApi itself? For example: `publicApi.__modules__`

Comment (rosen-vladimirov) [inline definitions/yok.d.ts]: sounds good and I think it will work :) 
I'll change it, thanks for the idea :)

**Comment by ligaz** [comment]
:+1:
